File: 20221129_Tuesday/20221129.rb
# Leetcode: 380. Insert Delete GetRandom O(2).
# https://leetcode.com/problems/insert-delete-getrandom-o1/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 368 ms, faster than 100.00% of Ruby online submissions for Insert Delete GetRandom O(1).
# Memory Usage: 248.5 MB, less than 100.00% of Ruby online submissions for Insert Delete GetRandom O(1).
# 2022.11.29 Daily Challenge.
# 2022.12.01 Updated.
class RandomizedSet
    def initialize()
      @h = {}
      @a = []
    end
=begin
    :type val: Integer
    :rtype: Boolean
=end
    def insert(val)
      return false if @h[val]
      @a.push(val)
      @h[val] = @a.size - 1
      return true
    end
=begin
    :type val: Integer
    :rtype: Boolean
=end
    def remove(val)
      return false if ! @h[val]
      # Swap in the array last element
      # and elemet to be delted.
      i,j = @h[val], @a.size - 1
      a,b = val,@a[j]
      @h[a],@h[b] = @h[b],@h[a]
      @a[i],@a[j] = @a[j],@a[i]
      @a.pop
      @h[val] = nil
      return true
    end


=begin
    :rtype: Integer
=end
    def get_random()
      @a.sample
    end
end

# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet.new()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.get_random()

File: 20221129_Tuesday/20221129.go
type RandomizedSet struct {
  n int
  a []int
  h map[int]int
}


func Constructor() RandomizedSet {
  return RandomizedSet{0, make([]int,0), make(map[int]int));
    
}


func (this *RandomizedSet) Insert(val int) bool {
  if _, ok := this.h[val]; ! ok {
    return false
  }
  this.a = append(this.a,val)
  this.h[val] = this.n 
  this.n += 1
  return true
}


func (this *RandomizedSet) Remove(val int) bool {
  if i, ok := this.h[val]; ! ok {
    return false
  }
  // Write last at the deleted place.
  last := this.a[this.n-1]
  this.h[last] = i
  delete(this.h, val)
  this.a = this.a[:this.n-1]
  this.n -= 1
  return true
}


func (this *RandomizedSet) GetRandom() int {
    
}


/**
 * Your RandomizedSet object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Insert(val);
 * param_2 := obj.Remove(val);
 * param_3 := obj.GetRandom();
 */
